### PR TITLE
Adding rigidbodymotion to grains

### LINKED ISSUE #5279: Add rigid body motion of grains in phase field module
We could add translation and rotation components for grains in phase field module. This will help us taking care of grain movements in porous media or due to gravity.

### PR DESCRIPTION
**Pull request by SudiptaBiswas**
ComputeGrainCenterUserObjects: Calculates volume and co-ordinate of the center of grains
ComputeGrainForceAndTorque: Userobject to calculate force acting on a grain
ConstatntGrainForceAndTorque: UserObjcet for providing constant force and torque applied to grains
GrainForceAndTorqueInterface: Userobject interface for referring/executing any of above two userobjects

GrainCentersPostProcessor: VectorPostprocessor to output grain center and volume from ComputeGrainCenterUserobject
GrainForcesPostProcessor: VectorPostprocessor to output forces and torques acting on grain from UserObject

ForceDensityMaterial: Calculates reaction forces due to interaction between grains
GrainAdvectionVelocity: Material to calculate advection velocity acting on a grain 

MultigrainRigidBodyMotion:  Additional advection kernel corresponding to  CH equation 
SingleGrainRigidBodyMotion: Additional advection kernel corresponding to  AC equation 

Closes #5279.

### CODE REVIEW

File: modules/phase_field/include/userobjects/ComputeGrainCenterUserObject.h [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+  const std::vector<Real> & getGrainVolumes() const;
+  const std::vector<Point> & getGrainCenters() const;
+
+protected:
+  unsigned int _qp;
+  unsigned int _ncrys;
+  std::vector<VariableValue *> _vals;
+  unsigned int _ncomp;
+  /// storing volumes and centers of all the grains
```
Comment (dschwen) [inline modules/phase_field/include/userobjects/ComputeGrainCenterUserObject.h]: wrap this in
```
///@{
...
///@}
```

File: modules/phase_field/include/materials/ForceDensityMaterial.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
+#ifndef FORCEDENSITYMATERIAL_H
```
Comment (dschwen) [inline modules/phase_field/include/materials/ForceDensityMaterial.h]: This needs a header

File: modules/phase_field/include/userobjects/ComputeGrainCenterUserObject.h [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+/*           (c) 2010 Battelle Energy Alliance, LLC             */
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
```
Comment (dschwen) [inline modules/phase_field/include/userobjects/ComputeGrainCenterUserObject.h]: Please use the modules header (like in the files above). And no empty lines after the header.

File: modules/phase_field/include/userobjects/ComputeGrainForceAndTorque.h [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+
```
Comment (dschwen) [inline modules/phase_field/include/userobjects/ComputeGrainForceAndTorque.h]: Same here, different header, no empty lines.

File: modules/phase_field/include/userobjects/ComputeGrainForceAndTorque.h [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+#ifndef COMPUTEGRAINFORCEANDTORQUE_H
+#define COMPUTEGRAINFORCEANDTORQUE_H
+
+#include "GrainForceAndTorqueInterface.h"
+#include "ElementUserObject.h"
+#include "ComputeGrainCenterUserObject.h"
```
Comment (dschwen) [inline modules/phase_field/include/userobjects/ComputeGrainForceAndTorque.h]: You can forward-declare this class using
```
class ComputeGrainCenterUserObject;
```
and move the include into the.C file.

File: modules/phase_field/include/userobjects/ComputeGrainForceAndTorque.h [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+  const ComputeGrainCenterUserObject & _grain_data;
+  const std::vector<Real> & _grain_volumes;
+  const std::vector<Point> & _grain_centers;
+  unsigned int _ncrys;
+  unsigned int _ncomp;
+  std::vector<RealGradient> _force_values;
+  std::vector<RealGradient> _torque_values;
+  std::vector<RealGradient> _force_derivatives;
+  std::vector<RealGradient> _torque_derivatives;
+  /// vector storing grain force and torque values
```
Comment (dschwen) [inline modules/phase_field/include/userobjects/ComputeGrainForceAndTorque.h]: The comments don't make it clear what the difference between the _store and _values variables is. Could you clarify this? I think one is a temporary storage for parallel communication and the other one is the final copy.

File: modules/phase_field/include/userobjects/ConstantGrainForceAndTorque.h [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+
```
Comment (dschwen) [inline modules/phase_field/include/userobjects/ConstantGrainForceAndTorque.h]: Different header, no empty lines

File: modules/phase_field/include/userobjects/ConstantGrainForceAndTorque.h [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+  std::vector<Real> _M;
+
+  unsigned int _ncrys;
+  unsigned int _ncomp;
+  std::vector<RealGradient> _force_values;
+  std::vector<RealGradient> _torque_values;
+  std::vector<RealGradient> _force_derivatives;
+  std::vector<RealGradient> _torque_derivatives;
+  /// vector storing grain force and torque values
+  std::vector<Real> _force_torque_store;
```
Comment (dschwen) [inline modules/phase_field/include/userobjects/ConstantGrainForceAndTorque.h]: You shouldn't be needing the *_store variables in this class, right?

File: modules/phase_field/include/userobjects/GrainForceAndTorqueInterface.h [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+
```
Comment (dschwen) [inline modules/phase_field/include/userobjects/GrainForceAndTorqueInterface.h]: modules header, no empty lines

File: modules/phase_field/include/vectorpostprocessors/GrainCentersPostprocessor.h [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+class GrainCentersPostprocessor;
+
+template<>
+InputParameters validParams<GrainCentersPostprocessor>();
+
+/**
+ *  GrainCentersPostprocessor is a type of VectorPostprocessor that outputs center and volume of grains
+ *  calculated in GrainCenterUserObject.
+ */
+
```
Comment (dschwen) [inline modules/phase_field/include/vectorpostprocessors/GrainCentersPostprocessor.h]: no empty line after the comment block

File: modules/phase_field/include/vectorpostprocessors/GrainCentersPostprocessor.h [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+/*           (c) 2010 Battelle Energy Alliance, LLC             */
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
```
Comment (dschwen) [inline modules/phase_field/include/vectorpostprocessors/GrainCentersPostprocessor.h]: header

File: modules/phase_field/include/vectorpostprocessors/GrainCentersPostprocessor.h [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+public:
+  /**
+    * Class constructor
+    * @param name The name of the object
+    * @param parameters The input parameters
+    */
+  GrainCentersPostprocessor(const InputParameters & parameters);
+
+  /**
+   * Destructor
```
Comment (dschwen) [inline modules/phase_field/include/vectorpostprocessors/GrainCentersPostprocessor.h]: That comment is a bit too obvious ;-). Let's remove it.

File: modules/phase_field/include/userobjects/ComputeGrainForceAndTorque.h (hunk 6 above)
Comment (SudiptaBiswas) [inline modules/phase_field/include/userobjects/ComputeGrainForceAndTorque.h]: Yes, the _store is used only for calculation purpose and for parallel communication.  _values are for providing final values that can be used by other objects.

File: modules/phase_field/include/vectorpostprocessors/GrainCentersPostprocessor.h [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+   */
+  virtual ~GrainCentersPostprocessor() {}
+
+  /**
+   * Initialize, clears the postprocessor vector
```
Comment (dschwen) [inline modules/phase_field/include/vectorpostprocessors/GrainCentersPostprocessor.h]: Let's remove this one as well (the method is empty anyways)

File: modules/phase_field/include/vectorpostprocessors/GrainCentersPostprocessor.h [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+#ifndef GRAINCENTERSPOSTPROCESSOR_H
+#define GRAINCENTERSPOSTPROCESSOR_H
+
+#include "GeneralVectorPostprocessor.h"
+#include "ComputeGrainCenterUserObject.h"
```
Comment (dschwen) [inline modules/phase_field/include/vectorpostprocessors/GrainCentersPostprocessor.h]: Again I would go with a forward-declare and move include to the.C file.

File: modules/phase_field/include/vectorpostprocessors/GrainCentersPostprocessor.h [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+   */
+  virtual ~GrainCentersPostprocessor() {}
+
+  /**
+   * Initialize, clears the postprocessor vector
+   */
+  virtual void initialize(){};
+
+  /**
+   * Populates the postprocessor vector of values with the supplied postprocessors
```
Comment (dschwen) [inline modules/phase_field/include/vectorpostprocessors/GrainCentersPostprocessor.h]: Please use ```///``` for single line doc comments

File: modules/phase_field/include/vectorpostprocessors/GrainForcesPostprocessor.h [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+class GrainForcesPostprocessor;
+
+template<>
+InputParameters validParams<GrainForcesPostprocessor>();
+
+/**
+ *  GrainForcesPostprocessor is a type of VectorPostprocessor that outputs the
+ *  force and torque values calculated in UserObjects.
+ */
+
```
Comment (dschwen) [inline modules/phase_field/include/vectorpostprocessors/GrainForcesPostprocessor.h]: All comments from ```GrainCentersPostprocessor``` apply to this file as well.

File: modules/phase_field/src/base/PhaseFieldApp.C [hunk 17]
```diff
@@ -36,10 +36,12 @@
 #include "LangevinNoise.h"
 #include "MaskedBodyForce.h"
 #include "MatDiffusion.h"
+#include "MultiGrainRigidBodyMotion.h"
 #include "PFFracBulkRate.h"
 #include "PFFracCoupledInterface.h"
 #include "PFFracIntVar.h"
 #include "SoretDiffusion.h"
+#include "SingleGrainRigidBodyMotion.h"
```
Comment (dschwen) [inline modules/phase_field/src/base/PhaseFieldApp.C]: This is not in alphabetical order.

File: modules/phase_field/src/base/PhaseFieldApp.C [hunk 18]
```diff
@@ -303,14 +325,11 @@ PhaseFieldApp::registerObjects(Factory & factory)
...
+  registerVectorPostprocessor(GrainForcesPostprocessor);
+
   registerMesh(EBSDMesh);
   registerMesh(ImageMesh);
-
-#undef registerDeprecatedObjectName
-#define registerDeprecatedObjectName(obj, name, time) factory.regLegacyReplaced<obj>(stringifyName(obj), name, time)
-#undef registerObject
-#define registerObject(name) factory.regLegacy<name>(stringifyName(name))
-
```
Comment (dschwen) [inline modules/phase_field/src/base/PhaseFieldApp.C]: Why did you remove this block? Who added it?

File: modules/phase_field/src/kernels/MultiGrainRigidBodyMotion.C [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+/****************************************************************/
+#include "MultiGrainRigidBodyMotion.h"
+
+template<>
+InputParameters validParams<MultiGrainRigidBodyMotion>()
+{
+  InputParameters params = validParams<Kernel>();
+  params.addClassDescription("Adds rigid mody motion to grains");
+  params.addRequiredCoupledVar("c", "Concentration");
+  params.addParam<MaterialPropertyName>("advection_velocity", "Material property for advection velocities");
```
Comment (dschwen) [inline modules/phase_field/src/kernels/MultiGrainRigidBodyMotion.C]: You need to make thos params either required, or provide a default, or check if they are valid before you use them.
My preference is providing a default value! And make the defaults consistent across the classes to that the users won't have to specify the parameter at all in most cases.

File: modules/phase_field/include/userobjects/ConstantGrainForceAndTorque.h (hunk 8 above)
Comment (SudiptaBiswas) [inline modules/phase_field/include/userobjects/ConstantGrainForceAndTorque.h]: Yes, those are redundant here. Will remove them.

File: modules/phase_field/src/kernels/MultiGrainRigidBodyMotion.C [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+MultiGrainRigidBodyMotion::MultiGrainRigidBodyMotion(const InputParameters & parameters) :
+    Kernel(parameters),
+    _c_var(coupled("c")),
+    _c(coupledValue("c")),
+    _grad_c(coupledGradient("c")),
+    _velocity_advection(getMaterialProperty<std::vector<RealGradient> >("advection_velocity")),
+    _div_velocity_advection(getMaterialProperty<std::vector<Real> >("advection_velocity_divergence")),
+    _velocity_advection_derivative(getMaterialProperty<std::vector<RealGradient> >("advection_velocity_derivative")),
```
Comment (dschwen) [inline modules/phase_field/src/kernels/MultiGrainRigidBodyMotion.C]: The *_derivative parameters are not declared in validParams!

File: modules/phase_field/src/kernels/SingleGrainRigidBodyMotion.C [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+  InputParameters params = validParams<Kernel>();
+  params.addClassDescription("Adds rigid mody motion to a single grain");
+  params.addRequiredCoupledVar("c", "Concentration");
+  params.addParam<unsigned int>("op_index",0, "Grain number for the kernel to be applied");
+  params.addParam<MaterialPropertyName>("advection_velocity", "Material property giving the advection velocity of grains");
+  params.addParam<MaterialPropertyName>("advection_velocity_divergence", "Material property for divergence of advection velocities");
+  return params;
+}
+
+SingleGrainRigidBodyMotion::SingleGrainRigidBodyMotion(const InputParameters & parameters) :
```
Comment (dschwen) [inline modules/phase_field/src/kernels/SingleGrainRigidBodyMotion.C]: ```SingleGrainRigidBodyMotion``` and ```MultiGrainRigidBodyMotion``` should have a common base class ```GrainRigidBodyMotionBase```. They have a lot in common, like most member variables and parameters.

File: modules/phase_field/src/kernels/SingleGrainRigidBodyMotion.C [hunk 22]
```diff
@@ -0,0 +1,62 @@
...
+  if (_c_var == jvar) //Requires c jacobian
+    return computeQpCJacobian();
+
+  return 0.0;
+}
+
+Real
+SingleGrainRigidBodyMotion::computeQpCJacobian()
+{
+  return _velocity_advection[_qp][_op_index] * _grad_phi[_j][_qp] * _test[_i][_qp] + _velocity_advection_derivative[_qp][_op_index] * _grad_c[_qp] * _phi[_j][_qp] *  _test[_i][_qp] + _div_velocity_advection[_qp][_op_index] * _phi[_j][_qp] * _test[_i][_qp] + _div_velocity_advection_derivative[_qp][_op_index] * _c[_qp] * _phi[_j][_qp] * _test[_i][_qp];
```
Comment (dschwen) [inline modules/phase_field/src/kernels/SingleGrainRigidBodyMotion.C]: Please break long equations into multiple lines.

File: modules/phase_field/src/base/PhaseFieldApp.C (hunk 18 above)
Comment (SudiptaBiswas) [inline modules/phase_field/src/base/PhaseFieldApp.C]: I think this was showing some conflict when I updated moose and that's why I removed these lines.

File: modules/phase_field/src/materials/ForceDensityMaterial.C [hunk 23]
```diff
@@ -0,0 +1,61 @@
...
+
+template<>
+InputParameters validParams<ForceDensityMaterial>()
+{
+  InputParameters params = validParams<Material>();
+  params.addClassDescription("Calculating the force density acting on a grain");
+  params.addCoupledVar("etas", "Array of coupled order parameters");
+  params.addCoupledVar("c", "Concentration field");
+  params.addParam<Real>("ceq",0.9816, "Equilibrium density");
+  params.addParam<Real>("cgb",0.25, "Thresold Concentration for GB");
```
Comment (dschwen) [inline modules/phase_field/src/materials/ForceDensityMaterial.C]: space after ,

File: modules/phase_field/src/materials/ForceDensityMaterial.C [hunk 24]
```diff
@@ -0,0 +1,61 @@
...
+  return params;
+}
+
+ForceDensityMaterial::ForceDensityMaterial(const InputParameters & parameters) :
+   Material(parameters),
+   _c(coupledValue("c")),
+   _ceq(getParam<Real>("ceq")),
+   _cgb(getParam<Real>("cgb")),
+   _k(getParam<Real>("k")),
+   _ncrys(coupledComponents("etas")), //determine number of grains from the number of names passed in.  Note this is the actual number -1
```
Comment (dschwen) [inline modules/phase_field/src/materials/ForceDensityMaterial.C]: Really? Why -1?

File: modules/phase_field/src/materials/ForceDensityMaterial.C [hunk 25]
```diff
@@ -0,0 +1,61 @@
...
+    for (unsigned int j = 0; j < _ncrys; ++j)
+    {
+      if (j != i)
+      {
+        product_etas[i] = (*_vals[i])[_qp] * (*_vals[j])[_qp] >= _cgb? 1 : 0; //Sum all other order parameters
+        sum_grad_etas[i] += product_etas[i] * ((*_grad_vals[i])[_qp] - (*_grad_vals[j])[_qp]);
+      }
+    }
+
+   _dF[_qp][i] = _k * (_c[_qp] - _ceq) * sum_grad_etas[i];
```
Comment (dschwen) [inline modules/phase_field/src/materials/ForceDensityMaterial.C]: Indentation is off by one

File: modules/phase_field/src/materials/ForceDensityMaterial.C (hunk 24 above)
Comment (SudiptaBiswas) [inline modules/phase_field/src/materials/ForceDensityMaterial.C]: Oh no it isn't. I missed modifying the comment.

File: modules/phase_field/src/materials/GrainAdvectionVelocity.C [hunk 26]
```diff
@@ -0,0 +1,79 @@
...
+template<>
+InputParameters validParams<GrainAdvectionVelocity>()
+{
+  InputParameters params = validParams<Material>();
+  params.addClassDescription("Calculation the advection velocity of grain due to rigid vody translation and rotation");
+  params.addCoupledVar("etas", "Array of other coupled order parameters");
+  params.addParam<Real>("translation_constant",500, "constant value characterizing grain translation");
+  params.addParam<Real>("rotation_constant",1.0, "constant value characterizing grain rotation");
+  params.addParam<UserObjectName>("grain_data","userobject for getting volume and center of mass of grains");
+  params.addParam<UserObjectName>("grain_force","userobject for getting force and torque acting on grains");
```
Comment (dschwen) [inline modules/phase_field/src/materials/GrainAdvectionVelocity.C]: some more lines with no space after,

File: modules/phase_field/src/materials/ForceDensityMaterial.C [hunk 27]
```diff
@@ -0,0 +1,61 @@
...
+  {
+    _vals[i] = &coupledValue("etas", i);
+    _grad_vals[i] = &coupledGradient("etas", i);
+  }
+}
+
+void
+ForceDensityMaterial::computeQpProperties()
+{
+  std::vector<Real> product_etas(_ncrys);
```
Comment (dschwen) [inline modules/phase_field/src/materials/ForceDensityMaterial.C]: I'd like to see those as member variables. Building an std::vector for each call to computeQpProperties is pretty slow!

File: modules/phase_field/src/materials/GrainAdvectionVelocity.C [hunk 28]
```diff
@@ -0,0 +1,79 @@
...
+  Real div_velocity_rotation_derivative;
+
+  _velocity_advection[_qp].resize(_ncrys);
+  _div_velocity_advection[_qp].resize(_ncrys);
+  _velocity_advection_derivative[_qp].resize(_ncrys);
+  _div_velocity_advection_derivative[_qp].resize(_ncrys);
+
+  for (unsigned int i = 0; i < _ncrys; ++i)
+  {
+    velocity_translation = _mt / _grain_volumes[i] * ((*_vals[i])[_qp] * _grain_forces[i]);
```
Comment (dschwen) [inline modules/phase_field/src/materials/GrainAdvectionVelocity.C]: It might be beneficial to declare these variables down here ast ```const RealGradient``` rather than declaring them above. This makes it clearer that the variables are only used as aliases for the terms and are not modified after their initial definition.

File: modules/phase_field/src/userobjects/ComputeGrainCenterUserObject.C [hunk 29]
```diff
@@ -0,0 +1,91 @@
...
+    ElementUserObject(parameters),
+    _ncrys(coupledComponents("etas")), //determine number of grains from the number of names passed in.  Note this is the actual number -1
+    _vals(_ncrys), //Size variable arrays
+    _ncomp(4*_ncrys),
+    _grain_data(_ncomp),
+    _grain_volumes(_ncrys),
+    _grain_centers(_ncrys)
+{
+  for (unsigned int i = 0; i < _ncrys; ++i)
+    _vals[i] = &coupledValue("etas",i);
```
Comment (dschwen) [inline modules/phase_field/src/userobjects/ComputeGrainCenterUserObject.C]: space after ,

File: modules/phase_field/include/vectorpostprocessors/GrainCentersPostprocessor.h (hunk 12 above)
Comment (SudiptaBiswas) [inline modules/phase_field/include/vectorpostprocessors/GrainCentersPostprocessor.h]: Do you want only the '// Destructor' comment to be removed or all of these ?

File: modules/phase_field/src/userobjects/ComputeGrainCenterUserObject.C [hunk 30]
```diff
@@ -0,0 +1,91 @@
...
+/*           (c) 2010 Battelle Energy Alliance, LLC             */
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
```
Comment (dschwen) [inline modules/phase_field/src/userobjects/ComputeGrainCenterUserObject.C]: More headers (please use the modules header)

File: modules/phase_field/include/vectorpostprocessors/GrainCentersPostprocessor.h (hunk 12 above)
Comment (dschwen) [inline modules/phase_field/include/vectorpostprocessors/GrainCentersPostprocessor.h]: Let's remove all "boilerplate" comments and retain only the ones that are
specific to your classes.

File: modules/phase_field/src/userobjects/ComputeGrainForceAndTorque.C [hunk 31]
```diff
@@ -0,0 +1,140 @@
...
+/****************************************************************/
+
+#include "ComputeGrainForceAndTorque.h"
+
+template<>
+InputParameters validParams<ComputeGrainForceAndTorque>()
+{
+  InputParameters params = validParams<ElementUserObject>();
+  params.addClassDescription("Userobject for calculating force and torque acting on a grain");
+  params.addRequiredParam<MaterialPropertyName>("force_density", "Force density material");
```
Comment (dschwen) [inline modules/phase_field/src/userobjects/ComputeGrainForceAndTorque.C]: I'd go for providing a default value here, rather than making it required.

File: modules/phase_field/tests/rigidbodymotion/grain_advection_constforce.i [hunk 32]
```diff
@@ -0,0 +1,192 @@
+#
+# Test the split parsed function free enery Cahn-Hilliard Bulk kernel
```
Comment (dschwen) [inline modules/phase_field/tests/rigidbodymotion/grain_advection_constforce.i]: Ha, these comments should all be changed! This is the problem with having too many comments, if you copy paste or change code you need to make sure to check if the comments are still valid :-)

File: modules/phase_field/include/vectorpostprocessors/GrainCentersPostprocessor.h (hunk 12 above)
Comment (SudiptaBiswas) [inline modules/phase_field/include/vectorpostprocessors/GrainCentersPostprocessor.h]: Okay.

File: modules/phase_field/examples/rigidbodymotion/AC_CH_advection_constforce_rect.i [hunk 33]
```diff
@@ -0,0 +1,213 @@
+[Mesh]
```
Comment (dschwen) [inline modules/phase_field/examples/rigidbodymotion/AC_CH_advection_constforce_rect.i]: Could you put *a lot more* comments into this example file please. Every block relating to rbm should have a few lines explaining what it is doing.

**Comment by dschwen** [comment]
Apart from a few comments above this looks great.

File: modules/phase_field/src/base/PhaseFieldApp.C (hunk 18 above)
Comment (dschwen) [inline modules/phase_field/src/base/PhaseFieldApp.C]: Check the latest devel, and copy the lines back into your PR. Preferable squash that into the commit that removed them!

File: modules/phase_field/tests/rigidbodymotion/grain_advection_constforce.i (hunk 32 above)
Comment (SudiptaBiswas) [inline modules/phase_field/tests/rigidbodymotion/grain_advection_constforce.i]: Yes, I should be more careful about the comments.

File: modules/phase_field/src/base/PhaseFieldApp.C (hunk 18 above)
Comment (SudiptaBiswas) [inline modules/phase_field/src/base/PhaseFieldApp.C]: okay

File: modules/phase_field/src/materials/ForceDensityMaterial.C (hunk 27 above)
Comment (SudiptaBiswas) [inline modules/phase_field/src/materials/ForceDensityMaterial.C]: I wanted to make these local variables that's why kept it here. Will move these.

**Comment by dschwen** [comment]
You have a merge conflict in ```PhaseFieldApp.C```. Check out the ```merge-conflicts``` package for Atom!

File: modules/phase_field/src/kernels/MultiGrainRigidBodyMotion.C [hunk 34]
```diff
@@ -0,0 +1,71 @@
...
+MultiGrainRigidBodyMotion::computeQpOffDiagJacobian(unsigned int jvar)
+{
+  if (_c_var == jvar) //Requires c jacobian
+    return computeQpCJacobian();
+
+  return 0.0;
+}
+
+Real
+MultiGrainRigidBodyMotion::computeQpCJacobian()
```
Comment (permcody) [inline modules/phase_field/src/kernels/MultiGrainRigidBodyMotion.C]: What is computeQp_C_Jacobian?

File: modules/phase_field/include/kernels/MultiGrainRigidBodyMotion.h [hunk 35]
```diff
@@ -0,0 +1,30 @@
...
+class MultiGrainRigidBodyMotion : public GrainRigidBodyMotionBase
+{
+public:
+  MultiGrainRigidBodyMotion(const InputParameters & parameters);
+
+protected:
+  virtual Real computeQpResidual();
+  virtual Real computeQpJacobian();
+  virtual Real computeQpOffDiagJacobian(unsigned int jvar);
+  virtual Real computeQpCJacobian();
```
Comment (jwpeterson) [inline modules/phase_field/include/kernels/MultiGrainRigidBodyMotion.h]: I'm not a huge fan of this function name (which seems to have been originally introduced by @tonkmr in his `SoretDiffusion` class and emulated here) -- at first I thought it was some kind of typo...  I'd also like to know a bit more about why this needs to exist outside of the usual `computeQpOffDiagJacobian()` hierarchy?  If `jvar==_c_var`, why not just call `Base::computeQpOffDiagJacobian()`?

File: modules/phase_field/src/kernels/MultiGrainRigidBodyMotion.C (hunk 34 above)
Comment (SudiptaBiswas) [inline modules/phase_field/src/kernels/MultiGrainRigidBodyMotion.C]: This is something introduced to calculate the jacobian for variable C. This is introduced so that the same kernel can be used by split and non-split version of CH kernels.

File: modules/phase_field/include/kernels/MultiGrainRigidBodyMotion.h (hunk 35 above)
Comment (SudiptaBiswas) [inline modules/phase_field/include/kernels/MultiGrainRigidBodyMotion.h]: Yes, I followed the same way as SoretDiffusion class. computeQpCJacobian() is called in computeQpOffDiagJacobian() as well as in computeQpJacobian().  This is introduced so that the same kernel can be used by split and non-split version of CH kernels.

Comment (dschwen) [inline modules/phase_field/include/kernels/MultiGrainRigidBodyMotion.h]: Yes, ```computeQpJacobian``` could defer to ```computeQpJacobian```, but I think factoring out the calculation of the Jacobian entry of ```c``` is somewhat more readable code IMO.

Comment (jwpeterson) [inline modules/phase_field/include/kernels/MultiGrainRigidBodyMotion.h]: I agree, but think that naming it so similarly to the "MOOSE" interface functions, not documenting it (maybe that is elsewhere), *and* placing it right under the MOOSE interface functions is confusing.

Comment (SudiptaBiswas) [inline modules/phase_field/include/kernels/MultiGrainRigidBodyMotion.h]: I can provide some comments explaining what is does, if that helps.

Comment (dschwen) [inline modules/phase_field/include/kernels/MultiGrainRigidBodyMotion.h]: Ok, let's find a better name then and separate it by an empty line. How about ```computeCVarJacobianEntry```?

**Comment by SudiptaBiswas** [comment]
See if 7d87420 looks better.

File: modules/phase_field/examples/rigidbodymotion/AC_CH_advection_constforce_rect.i [hunk 36]
```diff
@@ -0,0 +1,218 @@
...
+    type = SplitCHWRes
+    variable = w
+    mob_name = M
+  [../]
+  [./time]
+    type = CoupledTimeDerivative
+    variable = w
+    v = c
+  [../]
+  # advection kernel corrsponding to CH equation
```
Comment (dschwen) [inline modules/phase_field/examples/rigidbodymotion/AC_CH_advection_constforce_rect.i]: Can you put those comments inside the blocks? Otherwise Peacock will scramble them up.

File: modules/phase_field/src/kernels/GrainRigidBodyMotionBase.C [hunk 37]
```diff
@@ -0,0 +1,44 @@
...
+{
+  InputParameters params = validParams<Kernel>();
+  params.addClassDescription("Base class for adding rigid mody motion to grains");
+  params.addRequiredCoupledVar("c", "Concentration");
+  params.addRequiredCoupledVar("v", "Array of coupled variable names");
+  params.addParam<MaterialPropertyName>("advection_velocity", "advection_velocity", "Material property giving the advection velocity of grains");
+  params.addParam<MaterialPropertyName>("advection_velocity_divergence","advection_velocity_divergence", "Material property for divergence of advection velocities");
+  params.addParam<MaterialPropertyName>("advection_velocity_derivative_c", "advection_velocity_derivative_c", "Material property giving derivative of advection velocity of grains");
+  params.addParam<MaterialPropertyName>("advection_velocity_divergence_derivative_c", "advection_velocity_divergence_derivative_c", "Material property for derivative of divergence of advection velocities");
+    params.addParam<MaterialPropertyName>("advection_velocity_derivative_eta", "advection_velocity_derivative_eta", "Material property giving derivative of advection velocity of grains");
```
Comment (dschwen) [inline modules/phase_field/src/kernels/GrainRigidBodyMotionBase.C]: indentation

File: modules/phase_field/examples/rigidbodymotion/AC_CH_advection_constforce_rect.i (hunk 36 above)
Comment (SudiptaBiswas) [inline modules/phase_field/examples/rigidbodymotion/AC_CH_advection_constforce_rect.i]: Yeah Sure.

**Comment by dschwen** [comment]
You missed one last comment (header in include/userobjects/GrainForceAndTorqueInterface.h). Once that is resolved I think it can be merged.
